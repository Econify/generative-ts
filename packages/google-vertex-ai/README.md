# @generative-ts/google-vertex-ai

Google VertexAI support for [generative-ts](https://github.com/Econify/generative-ts/blob/main/README.md)
