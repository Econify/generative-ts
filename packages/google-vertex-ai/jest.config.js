const { pathsToModuleNameMapper } = require('ts-jest');
const { compilerOptions } = require('../../tsconfig.base.json');

module.exports = {
  displayName: 'google-vertex-ai',
  testEnvironment: 'node',
  rootDir: '../../',
  roots: ['<rootDir>/packages/google-vertex-ai/src'],
  transform: {
    '^.+\\.ts$': 'ts-jest'
  },
  testMatch: ['<rootDir>/packages/google-vertex-ai/src/**/*.spec.ts'],
  moduleFileExtensions: ['ts', 'js', 'json', 'node'],
  moduleNameMapper: pathsToModuleNameMapper(compilerOptions.paths, { prefix: '<rootDir>/' })
};
