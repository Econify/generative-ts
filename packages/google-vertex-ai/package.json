{
  "name": "@generative-ts/google-vertex-ai",
  "version": "0.1.0-alpha.6",
  "description": "Google Cloud VertexAI support for for generative-ts",
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "exports": {
    "default": "./dist/index.cjs",
    "import": "./dist/index.mjs",
    "types": "./dist/index.d.ts"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "scripts": {
    "build": "npm run clean && npm run build:rollup",
    "build:rollup": "rollup -c",
    "check": "npm run typecheck && npm run lint && npm run test",
    "clean": "rm -rf dist && rm -rf .rollup.cache && rm -f *.tsbuildinfo",
    "lint": "eslint src/",
    "lint:fix": "eslint --fix src/",
    "prepare": "npm run build",
    "prepublishOnly": "npm run check",
    "test": "jest",
    "test:cov": "jest --coverage",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "google-auth-library": "^9.10.0",
    "tslib": "^2.6.2"
  },
  "peerDependencies": {
    "@generative-ts/core": "^0.1.0-alpha.6"
  },
  "publishConfig": {
    "access": "public"
  },
  "license": "ISC",
  "repository": {
    "url": "Econify/generative-ts",
    "directory": "packages/google-vertex-ai"
  },
  "author": "John M Naglick",
  "homepage": "https://github.com/Econify/generative-ts#readme",
  "bugs": {
    "url": "https://github.com/Econify/generative-ts/issues"
  },
  "keywords": [
    "llm",
    "vertexai",
    "ts",
    "generative",
    "typescript"
  ]
}
