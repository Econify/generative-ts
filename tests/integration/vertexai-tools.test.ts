import { createVertexAiModelProvider } from "@packages/google-vertex-ai";

test("VertexAI - Google Gemini (Tools)", async () => {
  // arrange
  const model = await createVertexAiModelProvider({
    modelId: "gemini-1.0-pro",
  });

  // act
  const response = await model.sendRequest({
    system: "Use tools to help answer questions.",
    prompt: "What is the weather in Boston and New York City?",
    tools: [
      {
        function_declarations: [
          {
            name: "get_current_weather",
            description: "Get the current weather for a given location",
            parameters: {
              type: "OBJECT",
              properties: {
                city: {
                  type: "STRING",
                  description: "The city, e.g. San Francisco",
                },
                state: {
                  type: "STRING",
                  description: "The state, e.g. CA",
                },
              },
              required: ["city", "state"],
            },
          },
        ],
      },
    ],
  });

  // console.log(JSON.stringify(response.data.candidates, null, 2));

  // assert
  expect(response).toMatchApiSnapshot();
});

test("VertexAI - Google Gemini (Tools with Responses)", async () => {
  // arrange
  const model = await createVertexAiModelProvider({
    modelId: "gemini-1.0-pro",
  });

  // act
  const response = await model.sendRequest({
    system: "Use tools to help answer questions.",
    prompt: "",
    examplePairs: [
      {
        user: "When did the New York Mets win the World Series?",
        assistant:
          "{'answer': 'The New York Mets won the World Series in 1969 and 1986.'}",
      },
      {
        user: "Who is the best player in NY Mets history?",
        assistant: "{'answer': 'Tom Seaver'}",
      },
    ],
    contents: [
      {
        role: "user",
        parts: [
          {
            text: "What is the weather in Boston and New York City?",
          },
        ],
      },
      {
        role: "model",
        parts: [
          {
            function_call: {
              name: "get_current_weather",
              args: {
                city: "Boston",
                state: "MA",
              },
            },
          },
          {
            function_call: {
              name: "get_current_weather",
              args: {
                city: "New York City",
                state: "NY",
              },
            },
          },
        ],
      },
      {
        role: "user",
        parts: [
          {
            function_response: {
              name: "get_current_weather",
              response: {
                temperature: "70",
                temperature_unit: "F",
                pressure: "30.1",
                pressure_unit: "inHg",
                humidity: "50",
                condition: "Sunny",
              },
            },
          },
          {
            function_response: {
              name: "get_current_weather",
              response: {
                temperature: "22",
                temperature_unit: "F",
                pressure: "50.2",
                pressure_unit: "inHg",
                humidity: "55",
                condition: "Hurricanes",
              },
            },
          },
        ],
      },
    ],
    tools: [
      {
        function_declarations: [
          {
            name: "get_current_weather",
            description: "Get the current weather for a given location",
            parameters: {
              type: "OBJECT",
              properties: {
                city: {
                  type: "STRING",
                  description: "The city, e.g. San Francisco",
                },
                state: {
                  type: "STRING",
                  description: "The state, e.g. CA",
                },
              },
              required: ["city", "state"],
            },
          },
        ],
      },
    ],
  });

  // console.log(JSON.stringify(response.data.candidates, null, 2));

  // assert
  expect(response).toMatchApiSnapshot();
});
