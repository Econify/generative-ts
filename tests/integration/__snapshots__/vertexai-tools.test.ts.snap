// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VertexAI - Google Gemini ($tools) 1`] = `
{
  "config": {
    "body": "string",
    "errorRedactor": "function",
    "headers": {
      "Authorization": "string",
      "Content-Type": "string",
      "User-Agent": "string",
      "x-goog-api-client": "string",
    },
    "method": "string",
    "paramsSerializer": "function",
    "responseType": "string",
    "url": "string",
    "validateStatus": "function",
  },
  "data": {
    "candidates": [
      {
        "content": {
          "parts": [
            {
              "text": "string",
            },
          ],
          "role": "string",
        },
        "finishReason": "string",
        "safetyRatings": [
          {
            "category": "string",
            "probability": "string",
            "probabilityScore": "number",
            "severity": "string",
            "severityScore": "number",
          },
        ],
      },
    ],
    "usageMetadata": {
      "candidatesTokenCount": "number",
      "promptTokenCount": "number",
      "totalTokenCount": "number",
    },
  },
  "headers": {
    "alt-svc": "string",
    "cache-control": "string",
    "content-encoding": "string",
    "content-type": "string",
    "date": "string",
    "server": "string",
    "transfer-encoding": "string",
    "vary": "string",
    "x-content-type-options": "string",
    "x-frame-options": "string",
    "x-xss-protection": "string",
  },
  "request": {
    "responseURL": "string",
  },
  "status": "number",
  "statusText": "string",
}
`;

exports[`VertexAI - Google Gemini (COOL TOOLS) 1`] = `
{
  "config": {
    "body": "string",
    "errorRedactor": "function",
    "headers": {
      "Authorization": "string",
      "Content-Type": "string",
      "User-Agent": "string",
      "x-goog-api-client": "string",
    },
    "method": "string",
    "paramsSerializer": "function",
    "responseType": "string",
    "url": "string",
    "validateStatus": "function",
  },
  "data": {
    "candidates": [
      {
        "content": {
          "parts": [
            {
              "functionCall": {
                "args": {
                  "city": "string",
                  "state": "string",
                },
                "name": "string",
              },
            },
          ],
          "role": "string",
        },
        "finishReason": "string",
        "safetyRatings": [
          {
            "category": "string",
            "probability": "string",
            "probabilityScore": "number",
            "severity": "string",
            "severityScore": "number",
          },
        ],
      },
    ],
    "usageMetadata": {
      "candidatesTokenCount": "number",
      "promptTokenCount": "number",
      "totalTokenCount": "number",
    },
  },
  "headers": {
    "alt-svc": "string",
    "cache-control": "string",
    "content-encoding": "string",
    "content-type": "string",
    "date": "string",
    "server": "string",
    "transfer-encoding": "string",
    "vary": "string",
    "x-content-type-options": "string",
    "x-frame-options": "string",
    "x-xss-protection": "string",
  },
  "request": {
    "responseURL": "string",
  },
  "status": "number",
  "statusText": "string",
}
`;

exports[`VertexAI - Google Gemini (Tools with Responses) 1`] = `
{
  "config": {
    "body": "string",
    "errorRedactor": "function",
    "headers": {
      "Authorization": "string",
      "Content-Type": "string",
      "User-Agent": "string",
      "x-goog-api-client": "string",
    },
    "method": "string",
    "paramsSerializer": "function",
    "responseType": "string",
    "url": "string",
    "validateStatus": "function",
  },
  "data": {
    "candidates": [
      {
        "content": {
          "parts": [
            {
              "text": "string",
            },
          ],
          "role": "string",
        },
        "finishReason": "string",
        "safetyRatings": [
          {
            "category": "string",
            "probability": "string",
            "probabilityScore": "number",
            "severity": "string",
            "severityScore": "number",
          },
        ],
      },
    ],
    "usageMetadata": {
      "candidatesTokenCount": "number",
      "promptTokenCount": "number",
      "totalTokenCount": "number",
    },
  },
  "headers": {
    "alt-svc": "string",
    "cache-control": "string",
    "content-encoding": "string",
    "content-type": "string",
    "date": "string",
    "server": "string",
    "transfer-encoding": "string",
    "vary": "string",
    "x-content-type-options": "string",
    "x-frame-options": "string",
    "x-xss-protection": "string",
  },
  "request": {
    "responseURL": "string",
  },
  "status": "number",
  "statusText": "string",
}
`;

exports[`VertexAI - Google Gemini (Tools) 1`] = `
{
  "config": {
    "body": "string",
    "errorRedactor": "function",
    "headers": {
      "Authorization": "string",
      "Content-Type": "string",
      "User-Agent": "string",
      "x-goog-api-client": "string",
    },
    "method": "string",
    "paramsSerializer": "function",
    "responseType": "string",
    "url": "string",
    "validateStatus": "function",
  },
  "data": {
    "candidates": [
      {
        "content": {
          "parts": [
            {
              "functionCall": {
                "args": {
                  "city": "string",
                  "state": "string",
                },
                "name": "string",
              },
            },
          ],
          "role": "string",
        },
        "finishReason": "string",
        "safetyRatings": [
          {
            "category": "string",
            "probability": "string",
            "probabilityScore": "number",
            "severity": "string",
            "severityScore": "number",
          },
        ],
      },
    ],
    "usageMetadata": {
      "candidatesTokenCount": "number",
      "promptTokenCount": "number",
      "totalTokenCount": "number",
    },
  },
  "headers": {
    "alt-svc": "string",
    "cache-control": "string",
    "content-encoding": "string",
    "content-type": "string",
    "date": "string",
    "server": "string",
    "transfer-encoding": "string",
    "vary": "string",
    "x-content-type-options": "string",
    "x-frame-options": "string",
    "x-xss-protection": "string",
  },
  "request": {
    "responseURL": "string",
  },
  "status": "number",
  "statusText": "string",
}
`;
